### PR TITLE
Enable frakti to mix docker runtime with hyper runtime

### LINKED ISSUE #75: hybrid mode
I really like the idea of being able to schedule pods into lightweight vm's. There are times though when having access to the underlying node is needed. So its rather awkward to only have one at a time. It would be awesome if the docker backend and hyper backend could co'exist on the host at the same time, and pick one or the other based on pod annotation. You could then even possibly use kubernetes itself, with a daemonset that's docker based, to enable and manage hyper support on a kubernetes cluster. Looking at hyper a bit, it looks like it could containerize nicely.

### PR DESCRIPTION
**Pull request by resouer**
Fixes #75 

**Special Note:**
Also fixed bug: version of `apimachinery` pkg in current frakti is wrong, which is introduced by #73
The right way to fix this is making sure version of apimachinery is set by client-go (step 3):
1. (optional) under `$GOPATH/src/k8s.io/frakti`, `godep restore`
2. (optional) under `$GOPATH/src/k8s.io/kubernetes`, `godep restore`
3. under `$GOPATH/src/k8s.io/client-go`, `godep restore`
4. under `$GOPATH/src/k8s.io/frakti`, `godep save ./...`

Considering add this to godep save script in `hack`.

**Implementation details:**

Any pod: 
1. with annotation `runtime.frakti.alpha.kubernetes.io/OSContainer`
2. is configured to use host namespace (net, pid, ipc)

will be handled by build-in dockershim, which name is `alternative runtime`.

Special note for dockershim as alternative runtime:
1. individual streaming server is used for it.
2. CNI plugin will be re-used by alternative runtime, and it will fail if no CNI plugin provided (since in the future `kubenet` will be changed to CNI so we think this is OK).

Two tricks are included in this PR:
1. When restarted, frakti will have to load all containers from hyperd and dockerd to update alternative runtime cache. Please note, dockershim is able to filter out non-k8s managed containers.
2. Image operations (pull, list, remove) are executed parallel for two runtimes. So we only returns **Intersection** (suggested by @Crazykev ) of two runtimes as the view of frakti's image list.

I am not sure if we want to make alternative runtime optional (there's a TODO to address).
 
cc @feiskyer @gnawux

### CODE REVIEW

**Comment by feiskyer** [comment]
> Considering add this to godep save script in hack.

@resouer Open another PR for this?

File: cmd/frakti/frakti.go [hunk 1]
```diff
@@ -19,15 +19,29 @@ package main
...
+	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/kubelet/dockershim"
+	"k8s.io/kubernetes/pkg/kubelet/dockertools"
 	"k8s.io/kubernetes/pkg/kubelet/server/streaming"
 )
 
 const (
 	fraktiVersion = "0.1"
+	// TODO(resouer) frakti use cni by default, should we make this configurable?
```
Comment (feiskyer) [inline cmd/frakti/frakti.go]: I think don't, kubenet will also be replaced with cni in the future

File: cmd/frakti/frakti.go [hunk 2]
```diff
@@ -44,6 +58,8 @@ var (
 		"The directory for putting cni configuration file")
 	cniPluginDir = flag.String("cni-plugin-dir", "/opt/cni/bin",
 		"The directory for putting cni plugin binary file")
+	alternativeRuntimeEndpoint = flag.String("alternative-runtime", "unix:///var/run/docker.sock",
```
Comment (feiskyer) [inline cmd/frakti/frakti.go]: Maybe it's better to just name it as  `docker-endpoint` since it only works for docker.

File: cmd/frakti/frakti.go [hunk 3]
```diff
@@ -54,14 +70,26 @@ func main() {
...
 		fmt.Println("Initialize hyper runtime failed: ", err)
 		os.Exit(1)
 	}
 
-	server, err := manager.NewFraktiManager(hyperRuntime, hyperRuntime, streamingServer)
+	// 2. Initialize alternative runtime and start its own streaming server
+	alternativeStreamingConfig := getAlternativeStreamingConfig()
+	ds, err := initAlternativeRuntimeService(alternativeStreamingConfig)
+	if err != nil {
+		// TODO(harry) Do we want to make alternative runtime optional?
```
Comment (feiskyer) [inline cmd/frakti/frakti.go]: yes, it `alternativeRuntimeEndpoint` is not provided, we should disable it.

File: cmd/frakti/frakti.go [hunk 4]
```diff
@@ -70,14 +98,93 @@ func main() {
...
+		if err := httpServer.ListenAndServe(); err != nil {
+			fmt.Printf("Failed to start streaming server for alternative runtime: %v", err)
+			os.Exit(1)
+		}
+	}()
+}
+
+func initAlternativeRuntimeService(streamingConfig *streaming.Config) (dockershim.DockerService, error) {
+	// For now we use docker as the only supported alternative runtime
+	fmt.Printf("Initialize alternative runtime: docker runtime\n")
```
Comment (feiskyer) [inline cmd/frakti/frakti.go]: do not print by fmt, use log instead

File: pkg/manager/manager.go [hunk 5]
```diff
@@ -560,15 +561,52 @@ func (s *FraktiManager) Status(ctx context.Context, req *kubeapi.StatusRequest)
...
+	if firstError != nil {
+		glog.Error(firstError)
+		return nil, firstError
+	}
+
+	// NOTE: we show intersection of image list of hyper and alternative runtime
+	intersectList := getImageListIntersection(imageMapList[0], imageMapList[1])
+
+	// if there is different in two sides, print the different if log lever is high enough
+	if glog.V(5) && len(imageMapList[0]) != len(intersectList) {
```
Comment (feiskyer) [inline pkg/manager/manager.go]: what is `glog.V(5)` for?

File: pkg/manager/manager.go [hunk 6]
```diff
@@ -560,15 +561,52 @@ func (s *FraktiManager) Status(ctx context.Context, req *kubeapi.StatusRequest)
...
+	}
+
+	imageServiceList := []runtime.ImageService{s.imageService, s.alternativeRuntimeService}
+	serviceNameList := []string{"hyper image service", "alternative image service"}
+	imageMapList := make([]map[string]*kubeapi.Image, 2)
+
+	listImageFunc := func(i int) {
+		images, err := imageServiceList[i].ListImages(req.GetFilter())
+		if err != nil {
+			catchError(fmt.Errorf("ListImage from %s failed: %v", serviceNameList[i], err))
```
Comment (feiskyer) [inline pkg/manager/manager.go]: nit: a more readable way to handle errors https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/core/generic_scheduler.go#L182

File: pkg/manager/manager.go (hunk 5 above)
Comment (resouer) [inline pkg/manager/manager.go]: We assume `getImageListDifference` may require some time to compute, so we decide only calculate it only when glog.V(5) is meet, see:  https://github.com/golang/glog/blob/master/glog.go#L992

Comment (feiskyer) [inline pkg/manager/manager.go]: SGTM

Comment (Crazykev) [inline pkg/manager/manager.go]: compute difference everytime is expensive, just don't want it as default. `5` means higher than normal trace... Do you have any suggestion here?

**Comment by resouer** [comment]
CI failed due to https://github.com/kubernetes/kubernetes/issues/43267

**Comment by resouer** [comment]
I am adding e2e test for alternative runtime since we now have docker as dependency.

**Pull request merged**